File: backend/app/core/events/subscribers/task_broadcast.py
from __future__ import annotations

import json
import logging
from typing import Any, Optional

from app.core.events.envelope import ConductorEventEnvelope
from app.core.events.subscriber import SubscriberPhase
from app.core.sandbox_bridge import SandboxBridge, SandboxBridgeRegistry, WsOutMessage

logger = logging.getLogger(__name__)


class TaskBroadcastSubscriber:
    """Phase 2: broadcast events to task-level WebSocket subscribers and Redis."""

    name = "task_broadcast"
    phase = SubscriberPhase.BROADCAST

    def __init__(self, bridge_registry: SandboxBridgeRegistry) -> None:
        self._bridge_registry = bridge_registry

    async def handle(self, envelope: ConductorEventEnvelope) -> None:
        if envelope.task_id is None or envelope.task_broadcast_payload is None:
            return

        if envelope.sandbox_id is None:
            return

        bridge = await self._bridge_registry.get(envelope.sandbox_id)
        if bridge is None:
            return

        ws_msg = WsOutMessage(
            type="event",
            payload=envelope.task_broadcast_payload,
        )
        await bridge.broadcast_to_task(envelope.task_id, ws_msg)

        from app.core.lifespan import get_redis_coordinator
        coordinator = get_redis_coordinator()
        if coordinator:
            await coordinator.publish_event(
                envelope.task_id,
                json.dumps({"type": ws_msg.type, **ws_msg.payload}),
            )

File: backend/app/joysafeter_orchestrator_rs/src/events/task_broadcast.rs
use std::sync::Arc;

use tokio::sync::broadcast;
use tracing::warn;

use super::envelope::EventEnvelope;
use crate::kernel::ha::BridgeStore;

/// TaskBroadcastSubscriber — BROADCAST phase.
///
/// Fans out events to per-task WebSocket subscribers via SandboxBridge,
/// and publishes to Redis for cross-instance delivery.
/// Mirrors the Python `TaskBroadcastSubscriber`.
pub struct TaskBroadcastSubscriber {
    bridge_store: Arc<dyn BridgeStore>,
}

impl TaskBroadcastSubscriber {
    pub fn new(bridge_store: Arc<dyn BridgeStore>) -> Self {
        Self { bridge_store }
    }

    /// Spawn as a background task listening on the event bus.
    pub fn spawn(
        self,
        mut rx: broadcast::Receiver<Arc<EventEnvelope>>,
    ) -> tokio::task::JoinHandle<()> {
        tokio::spawn(async move {
            loop {
                match rx.recv().await {
                    Ok(envelope) => {
                        self.handle(&envelope).await;
                    }
                    Err(broadcast::error::RecvError::Lagged(n)) => {
                        warn!("TaskBroadcastSubscriber lagged by {n} messages");
                    }
                    Err(broadcast::error::RecvError::Closed) => break,
                }
            }
        })
    }

    async fn handle(&self, envelope: &EventEnvelope) {
        let task_id = match envelope.task_id {
            Some(id) => id,
            None => return,
        };

        let sandbox_id = match envelope.sandbox_id {
            Some(id) => id,
            None => return,
        };

        // Broadcast to per-task WebSocket subscribers via bridge
        if let Some(bridge) = self.bridge_store.get_by_db_id(sandbox_id) {
            let payload = envelope
                .task_broadcast_payload
                .as_ref()
                .unwrap_or(&envelope.payload);
            bridge.broadcast_to_task(task_id, payload.clone()).await;
        }
    }
}
